articles/WideStrings.md

```diff
@@ -0,0 +1,213 @@
+There is a lot of history here, and things are much more complicated then the brief introduction above.
+For more information, the following links provide introductory material:
+
+* [http://kunststube.net/encoding/](http://kunststube.net/encoding/)
+* [http://stackoverflow.com/questions/4588302/why-isnt-wchar-t-widely-used-in-code-for-linux-related-platforms](http://stackoverflow.com/questions/4588302/why-isnt-wchar-t-widely-used-in-code-for-linux-related-platforms)
+* [http://www.diveintopython3.net/strings.html](http://www.diveintopython3.net/strings.html)
+* [http://stackoverflow.com/questions/402283/stdwstring-vs-stdstring](http://stackoverflow.com/questions/402283/stdwstring-vs-stdstring)
+
+The rest of this document is generally going to assume Unicode, unless stated otherwise.
+
+## Unicode Characters in Strings
+ROS 1 says string fields are to contain ASCII encoded data, but allows UTF-8.
+DDS-XTYPES mandates UTF-8 be used as the encoding of IDL type `string`.
+To be compatible with ROS 1 and DDS-XTYPES, in ROS 2 the content of a `string` is expected to be UTF-8.
+
+## Introducing Wide Strings
+ROS 2 messages will have a new [primitive field type](/articles/interface_definition.html) `wstring`.
+This purpose is to allow ROS 2 nodes to comminicate with non-ROS DDS entities using an IDL containing a `wstring` field.
+The encoding of data in this type should be UTF-16 to match DDS-XTYPES 1.2.
+Since both UTF-8 and UTF-16 can encode the same code points, new ROS 2 messages should prefer `string` over `wstring`.
+
+## Encodings are Recommendations
+The choice of UTF-8 or UTF-16 for `string` and `wstring` are recommendations.
+A `string` or `wstring` field is allowed to be populated and published with an unknown encoding.
+If subscriber receives a string containing an unknown ncoding then it is up to the subscriber how to handle it.
+It could decode it using an encoding that has been agreed to out of band.
+Other subscribers like `ros2 topic echo` may echo the bytes in hexadecimal.
+
+In practice using different encodings may not be possible.
+The IDL specification forbids `string` from containing `NULL` values.
+For compatibility a ROS message `string` field must not contain zero bytes, and a `wstring` field must not contain zero words.
+An encoding like UTF-32 cannot be used since it may have zero bytes or words in a code point.
+If UTF-32 encoded strings need to be transmitted then one should either convert to UTF-8 and use a `string` or use a `uint32` array.
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> Can ROS 1 publish a string with NULL bytes?
+</div>
+
+## Unicode Strings Across ROS 1 Bridge
+Since ROS 1 and 2 both allow `string` to be UTF-8, the ROS 1 bridge will pass values unmodified between them.
+If a message sent from ROS 2 to ROS 1 fails to serialize because a string contains values that are not legal UTF-8 then it won't be published by the bridge.
```

Comment:
This might go away if we decide on a mandatory encoding. Anyway I don't think we should set in stone that such messages will be dropped. There are fall back behaviors like ignoring or replacing not-encodable characters which could be considered.

Reply:
It seems like in some cases replacing or dropping invalid characters could be unsafe. It is possible the string message is a command with an invalid character that would be changed by that strategy. Worst case it might turn a string like  `DELETE ■*` into `DELETE *` where ■ is standing in for an invalid code point.

Reply:
I changed the wording to say dropping the entire message is the default behavior while allowing one to opt-in to another strategy 4a21e094e0ea3ab7bc8beeeca67395a781bbf810